### DISCUSSION TITLE
Simple example to use Ztest for integration and unit testing

### DISCUSSION THREAD

**Discussion by MaazSk** [zephyrproject-rtos/zephyr#76526] [Q&A] [answered]
I am trying to use Ztest for unit and integration testing. I already have some of my application running I want to add the test folder to it which will have the test code but I am unsure how to do it. Is there a sample application which I could use with my current project.
Currently if I try to use Ztest it gives me error `multiple definition to main` so pretty much not sure how to start. Is there a proper guide with samples to use.

Thank you!

**Comment by koffes** [comment, marked as answer]
I would start looking at https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/test/ztest.rst
The most basic sample can be seen here: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/testsuite/integration/src/main.c

**Reply by MaazSk** [reply]
Yup I started with [this link](https://github.com/zephyrproject-rtos/zephyr/blob/main/doc/develop/test/ztest.rst) only. Now I want to add a custom board for testing but don't know how to do it. I am getting the following error when I run `west twister -p myBoard -T . --device-testing`
```
`INFO    - Using Ninja..
INFO    - Zephyr version: v3.7.0-rc3-1-gf028c76ab8b4
INFO    - Using 'zephyr' toolchain.
ERROR   - platform_filter - unrecognized platform - myBoard` 
```
I am able to do a `west build -p myBoard` and `west flash` but I get error when I try the twister thing

**Reply by koffes** [reply]
Why do you want to add your own board? The Ztests are usually run on some existing simulated hardware.
Ztests are (commonly) used for unit testing which is most valuable for modules which are less HW dependent. Of course, one can mock the HW.
Custom boards are mostly used when you want to add your own new physical hardware.

**Reply by MaazSk** [reply]
I am using stm32h750b_dk dev board but for some modifications for my project I create my own board based on stm32h750b_dk so all my project required board changes goes in the board that I created that is the reason I need to test it on custom board. Now I have made some progress but I get this error `FATAL ERROR: download file /home/name/projects/zep-workspace/application/tests/twister-out/stm32h750b_dk_myBoard/drivers.memc.stm32_sdram/zephyr/zephyr.signed.hex does not exist
`
I am using this command `west twister -T . --board-root=/home/name/projects/zep-workspace/zep-boards/ -p stm32h750b_dk_myBoard --device-testing --device-serial /dev/ttyACM0 --device-serial-baud 115200`

**Reply by koffes** [reply]
Thanks, that clarifies a bit.
I see that you are running tests on the actual HW with UART.

Looking at https://docs.zephyrproject.org/latest/develop/test/ztest.html it is stated: "The framework can be used in two ways, either as a generic framework for integration testing, or for unit testing specific modules."
I have mostly used it as the latter. 

For your unit testing, you should consider if the tests can be run in a simulator instead. See e.g.
https://docs.zephyrproject.org/latest/boards/qemu/cortex_a53/doc/index.html. That would greatly speed up as you don't have to flash the device and wait for a slow UART to transfer the data.
For Ztest integration tests running on HW, it has been some time since I did that and others may have to chime in.

**Reply by MaazSk** [reply]
Thank you! for the reply but I figured out and was able to use the UART of my board the issue was how I have set up some parameters in my `boards.cmake file` and making some tweaks it started working. 

Thanks a lot for the detailed answer.

**Reply by abdullahzaman** [reply]
@MaazSk I am having the same problem can u please share what modification you had to do?

**Reply by MaazSk** [reply]
@mustafaabdullahk what is the exact issue that you are facing?
